### ISSUE TITLE
[SR-13385] Improved diagnostics when using instance methods before initialization

### ISSUE THREAD

**Issue by swift-ci** [swiftlang/swift#55825]
|                  |                 |
   |------------------|-----------------|
   |Previous ID       | SR-13385      |
   |Radar             | rdar://problem/67361258         |
   |Original Reporter | srinikhil07 (JIRA User)      |
   |Type              | Improvement    |

   <details>
  <summary>Additional Detail from JIRA</summary>

   |                  |                 |
   |------------------|-----------------|
   |Votes             | 0         |
   |Component/s       | Compiler    |
   |Labels            | Improvement, DiagnosticsQoI, StarterBug        |
   |Assignee          | None      |
   |Priority          | Medium      |

   

   md5: de102270b8499a8c34615e7067e1329f

  </details>

**Issue Description:**

One of the most common errors beginners of Swift like me make is to start using class instances just by declaring but not initializing.

 

For example:  
 

``` java
var test : Dictionary<String, String>
 test.updateValue("1", forKey: "1")   
test["1"] = "1" // same error as below  
```

   
In this case, the compiler does emit the below error:  
*<span color="#de350b">variable 'test' passed by reference before being initialized</span>*  
   
I think we can redact "passed by reference" cos Swift guide <https://docs.swift.org/swift-book/LanguageGuide/Methods.html> doesn't explicitly state that internally we pass object by reference and this SO thread <https://stackoverflow.com/questions/24838015/variable-p-passed-by-reference-before-being-initialized> gives us an interesting example where the user keeps asking why the phrase "passed by reference" is used.

**Comment by stzn** [comment]
@LucianoPAlmeida 
Is it still active and can I try it as my first issue?

**Comment by AnthonyLatsis** [comment]
I agree that the "pass by reference" bit is conveying correct information at the expense of clarify. Either way the problem is indeed that one should initialise something before *using* it, however so.

**Comment by stzn** [comment]
Thank you for assigning me. I'll start working on this soon.

**Comment by stzn** [comment]
I have created the pull request. Please let me know if any additional actions are needed. Thank you.
